src/antpm-fit2gpx.1
.\" -*- coding: utf-8-unix -*-
.\" groff -man -Tutf8 antpm-downloader.1
.\" Copyright © 2013 by Kristóf Ralovich
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" License.
.\"
.TH ANTPM-FIT2GPX "1" "October 2013" "antpm-fit2gpx v1.9" "User Commands"
.SH NAME
antpm-fit2gpx \- console based FIT to GPX converter
.SH DESCRIPTION
.SS "Allowed options:"
.TP
\fB\-h\fR [ \fB\-\-help\fR ]
produce help message
.TP
\fB\-F\fR [ \fB\-\-fitFolder\fR ] arg
Folder with FIT files
.TP
\fB\-D\fR [ \fB\-\-decode\-fit\-root\fR ] arg
FIT file, encoding the root directory contents on a device, e.g. \fI/tmp/0000.fit\fP
.TP
\fB\-L\fR [ \fB\-\-fix\-last\-write\fR ]
Correct last\-written time of .fit files on disk.
No conversion to GPX takes place.
.TP
\fB\-\-v\fR arg
Adjust verbosity level, varies in [1, 6]
.TP
\fB\-V\fR [ \fB\-\-version\fR ]
Print version information
.SH "AUTHOR"
.B antpm-downloader
was written by Kristóf Ralovich. This man page is Copyright ©
2010-2013 Kristóf Ralovich.
.SH "SEE ALSO"
.UR "https://github.com/ralovich/antpm"
.BR "https://github.com/ralovich/antpm"
.UE


debian/changelog
antpm (1.8-1) UNRELEASED; urgency=low

  * GarminConvert: add a method that offsets the timestamps
  * AntMessage: bugfix: correctly convert file creation time

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:46:18 +0200

antpm (1.7-1) UNRELEASED; urgency=low

  * antpm-*: recongize -h option for --help too
  * fit: this is a run-time warning
  * docs: start writing user's guide
  * docs: how to submit a bug report
  * docs: add a note on installing the udev rule for user access to the
    USB2ANT stick
  * docs: refine udev procedure
  * docs: split maintainer documentation from end-user documentation
  * FIT: bugfix: don't offset timestamps twice
  * bugfix: this was a crippling bug
  * docs: add todo
  * new version

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:11:34 +0200

antpm (1.6-3) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * docs: more notes
  * this change was only necessary for win64
  * debian: refine c++ compiler dependency
  * serial: add more verbosity to output

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Wed, 26 Jun 2013 11:43:35 +0200

antpm (1.6-2) UNRELEASED; urgency=low

  [ Kristof Ralovich ]
  * script to generate changelog from git

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:10:28 +0200

antpm (1.6-0) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * include Release builds in a .deb. package
  * clarify build- and run-time dependencies
  * add todo
  * fix build with clang 3.0
  * add some documentation
  * fix build warning
  * git should ignore build folder
  * AntFr310XT2::createDownloadFolder()
  * further define folder name based on device S/N
  * cleanup
  * deb: lower build time dependencies
  * create download folder after knowing the device id
  * fit2gpx: option to decode ANT root folder
  * initial, file based logging
  * docs: note on performance
  * fix output
  * use centralized logging in more places
  * fix output
  * introduce namespace antpm
  * namespace conversion
  * missing namespace
  * more unified logging
  * more unified logging
  * cleanup
  * fix an assertion failure
  * unified logging replaced loggerc()
  * reindent
  * define NDEBUG in release
  * avoid crashes

  [ unknown ]
  * build fixes

  [ RALOVICH, Kristóf ]
  * add todo
  * add todo
  * move GARMIN_EPOCH to antdefs.hpp
  * bugfix: fix thinko
  * try setting date of .fit files
  * change logging
  * first stub for centralized per-device setting

  [ RALOVICH, Kristof ]
  * missing include
  * cmake: fix linking under win
  * previously the x86 dll hada weird name...
  * cmake: make sure the correct generator is invoked
  * fix 64-bit linking
  * cmake: fix determination of win64
  * scripts: build helper for win64
  * antpm-downloader: bugfix
  * update discriution list

  [ Kristof Ralovich ]
  * deb: relax build time dependencies

  [ RALOVICH, Kristof ]
  * deb: libusb was missing as a build-dep

  [ Kristof Ralovich ]
  * bugfix: resolve gcc-4.4 build issues

  [ RALOVICH, Kristof ]
  * pkgconfig should look for libusb-1.0

  [ Kristof Ralovich ]
  * cleanup
  * fix missing include
  * debian: fixing dependencies
  * account for boost:filesystem version changes

  [ RALOVICH, Kristof ]
  * cmake: fix finding libusb under nix

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:02:17 +0200

antpm (1.0-1) UNRELEASED; urgency=low

  * Initial release.

 -- Kristof Ralovich <tade60@freemail.hu>  Tue, 21 Mar 2013 16:22:10 +0100


src/antpm-fit2gpx.cpp
// -*- mode: c++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2; coding: utf-8-unix -*-
// ***** BEGIN LICENSE BLOCK *****
//////////////////////////////////////////////////////////////////////////
// Copyright (c) 2011-2014 RALOVICH, Kristóf                            //
//                                                                      //
// This program is free software; you can redistribute it and/or modify //
// it under the terms of the GNU General Public License as published by //
// the Free Software Foundation; either version 3 of the License, or    //
// (at your option) any later version.                                  //
//                                                                      //
// This program is distributed in the hope that it will be useful,      //
// but WITHOUT ANY WARRANTY; without even the implied warranty of       //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        //
// GNU General Public License for more details.                         //
//                                                                      //
//////////////////////////////////////////////////////////////////////////
// ***** END LICENSE BLOCK *****

#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>

#include "AntMessage.hpp"
#include "FIT.hpp"
#include "common.hpp"
#include "Log.hpp"

#include <boost/program_options.hpp>
#include <boost/exception/diagnostic_information.hpp>

namespace po = boost::program_options;
namespace fs = std::filesystem;
using namespace std;
using namespace antpm;


DEFAULT_LOG_INSTANTIATOR

const
std::vector<fs::path>
find_files(const fs::path & dir_path,         // in this directory,
           const std::string & ext // search for this extension,
           )
{
  std::vector<fs::path> paths;
  if ( !fs::exists( dir_path ) ) return paths;
  fs::directory_iterator end_itr; // default construction yields past-the-end
  for ( fs::directory_iterator itr( dir_path );
    itr != end_itr;
    ++itr )
  {
    if ( fs::is_directory(itr->status()) )
    {
      std::vector<fs::path> paths2(find_files( itr->path(), ext) );
      paths.insert(paths.end(), paths2.begin(), paths2.end());
    }
    else if ( fs::path(*itr).extension() == ext ) // see below
    {
      paths.push_back(*itr);
    }
  }
  return paths;
}



int
main(int argc, char** argv)
{
  antpm::Log::instance()->addSink(std::cout);
#ifndef NDEBUG
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_DBG);
#else
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_INF);
#endif

  // Declare the supported options.
  std::string fitFolder;
  std::string fitRootFile;
  bool        fixLastWrt = false;
  int         verbosityLevel = antpm::Log::instance()->getLogReportingLevel();
  po::options_description desc("Allowed options");
  std::vector<const char*> args(argv, argv+argc);
  po::variables_map vm;
  try
  {
    desc.add_options()
    ("help,h",                                                  "produce help message")
    ("fitFolder,F",      po::value<std::string>(&fitFolder),    "Folder with FIT files")
    ("decode-fit-root,D", po::value<std::string>(&fitRootFile), "FIT file, encoding the root directory contents on a device, e.g. /tmp/0000.fit")
    ("fix-last-write,L", po::value<bool>(&fixLastWrt)->zero_tokens()->implicit_value(true), "Correct last-written time of .fit files on disk. No conversion to GPX takes place.")
    ("v",                po::value<int>(&verbosityLevel),       "Adjust verbosity level, varies in [1, 6]")
    ("version,V",                                               "Print version information")
    ;

    //po::parsed_options parsed = po::parse_command_line(argc, argv, desc);
    po::parsed_options parsed = po::command_line_parser(argc, argv).options(desc).run();
    po::store(parsed, vm);
    po::notify(vm);
  }
  catch(po::error& error)
  {
    cerr << error.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }
  catch(boost::exception& e)
  {
    cerr << boost::diagnostic_information(e) << std::endl;
    return EXIT_FAILURE;
  }
  catch(std::exception& ex)
  {
    cerr << ex.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }

  if(vm.count("version") || vm.count("V"))
  {
    cout << argv[0] << " " << antpm::getVersionString() << "\n";
    return EXIT_SUCCESS;
  }

  if(vm.count("v"))
  {
    if(verbosityLevel >= antpm::LOG_ERR && verbosityLevel <= antpm::LOG_DBG3)
    {
      antpm::Log::instance()->setLogReportingLevel(static_cast<antpm::LogLevel>(verbosityLevel));
    }
  }

  if(vm.count("help") || vm.count("h")
     || (fitFolder.empty() && fitRootFile.empty()))
  {
    cout << desc << "\n";
    return EXIT_SUCCESS;
  }

  if(!fitRootFile.empty())
  {
    ZeroFileContent zfc;
    FIT fit;
    vector<uchar> v(readFile(fitRootFile.c_str()));
    if(fit.parseZeroFile(v, zfc))
       return EXIT_SUCCESS;
    else
      return EXIT_FAILURE;
  }

  if(!fs::is_directory(fs::path(fitFolder)))
  {
    LOG_VAR(fitFolder);
    CHECK_RETURN_RV_LOG_OK(fs::is_directory(fs::path(fitFolder)), EXIT_FAILURE);
  }

  std::vector<fs::path> fitFiles(find_files(fs::path(fitFolder), ".fit"));
  std::sort(fitFiles.begin(), fitFiles.end());

  for (size_t i = 0; i < fitFiles.size(); i++)
  {
    const std::string in(fitFiles[i].string());
    if(fixLastWrt) // for fixing up FIT file dates created before fixing time conversion bug
    {
      vector<uchar> fitBytes(readFile(in.c_str()));
      std::time_t fileCreationTime; // GMT/UTC
      bool has_ct = FIT::getCreationDate(fitBytes, fileCreationTime);
      if(has_ct)
      {
        time_t lastWrite = to_time_t(fs::last_write_time(fs::path(in))); // GMT/UTC

        char tbuf[256];
        strftime(tbuf, sizeof(tbuf), "%d-%m-%Y %H:%M:%S", localtime(&lastWrite));

        char tbuf2[256];
        strftime(tbuf2, sizeof(tbuf2), "%d-%m-%Y %H:%M:%S", localtime(&fileCreationTime));

        time_t diff = (lastWrite-fileCreationTime);
        if(diff>1 || diff<-1)
        {
          LOG(LOG_INF) << in << ", fixing "
               << " new=" << tbuf2
               << "  ---->  old=" << tbuf
               << "\n";
            fs::last_write_time(fs::path(in), from_time_t<fs::file_time_type>(fileCreationTime));
        }
        else
        {
          LOG(LOG_DBG) << "Last write time already correct: " << in << "\n";
        }
      }
      else
      {
        LOG(LOG_INF) << "No FIT creation time: " << in << "\n";
      }
    }
    else
    {
#if BOOST_VERSION<=104300
      const std::string out(fitFiles[i].parent_path().string()+"//"+fitFiles[i].stem()+".gpx");
#else
      const std::string out(fitFiles[i].parent_path().string()+"//"+fitFiles[i].stem().string()+".gpx");
#endif
      LOG_VAR2(in, out);
      GPX gpx;
      FIT fit;
      vector<uchar> v(readFile(in.c_str()));
      if(fit.parse(v, gpx))
        gpx.writeToFile(out);
    }
  }


  return EXIT_SUCCESS;
}
